src/fstext/fstext-utils-inl.h

```diff
@@ -114,6 +114,55 @@ void GetInputSymbols(const Fst<Arc> &fst,
   std::sort(symbols->begin(), symbols->end());
 }
 
+template<class Arc, class I>
+class RemoveSomeInputSymbolsMapper {
+public:
+  Arc operator ()(const Arc &arc_in) {
+    Arc ans = arc_in;
+    if (to_remove_set_.count(ans.ilabel) != 0) ans.ilabel = 0;  // remove this symbol
+    return ans;
+  }
+  MapFinalAction FinalAction() { return MAP_NO_SUPERFINAL; }
+  MapSymbolsAction InputSymbolsAction() { return MAP_CLEAR_SYMBOLS; }
+  MapSymbolsAction OutputSymbolsAction() { return MAP_COPY_SYMBOLS; }
+  uint64 Properties(uint64 props) const {
+    // remove the following as we don't know now if any of them are true.
+    uint64 to_remove = kAcceptor|kNotAcceptor|kIDeterministic|kNonIDeterministic|
+        kNoEpsilons|kNoIEpsilons|kILabelSorted|kNotILabelSorted;
+    return props & ~to_remove;
+  }
+  RemoveSomeInputSymbolsMapper(const std::vector<I> &to_remove):
+      to_remove_set_(to_remove) {
+    KALDI_ASSERT_IS_INTEGER_TYPE(I);
+         assert(to_remove_set_.count(0) == 0);  // makes no sense to remove epsilon.
+       }
+private:
+  kaldi::ConstIntegerSet<I> to_remove_set_;
+};
+
+template<class Arc, class I>
+using LookaheadFst = ArcMapFst<Arc, Arc, RemoveSomeInputSymbolsMapper<Arc, I> >;
+
+// Lookahead composition is used for optimized online
+// composition of FSTs during decoding. See
+// nnet3/nnet3-latgen-faster-lookahead.cc
+template<class Arc, class I>
+LookaheadFst<Arc, I> LookaheadComposeFst(const Fst<Arc> &ifst1,
+                                         const Fst<Arc> &ifst2,
+                                         const std::vector<I> &to_remove) {
+  typedef LookAheadMatcher< StdFst > M;
+  typedef AltSequenceComposeFilter<M> SF;
+  typedef LookAheadComposeFilter<SF, M>  LF;
+  typedef PushWeightsComposeFilter<LF, M> WF;
+  typedef PushLabelsComposeFilter<WF, M> ComposeFilter;
+
+  fst::CacheOptions cache_opts(true, 1 << 26LL);
+  ComposeFstOptions<StdArc, M, ComposeFilter> opts(cache_opts);
+
+  RemoveSomeInputSymbolsMapper<Arc, I> mapper(to_remove);
```

Comment:
I'm not sure why this mapper is necessary here... You're essentially constructing HCLG here. The input symbols are going to be HMM transition ids, right? We only need to use disambiguation symbols in the case of the lexicons' input symbols (phonemes), right? Can you explain the purpose of this? It seems like you may accidentally remove valid transition ID's accidentally as it is now.

Reply:
Without disambig symbols HCL is not deterministic anymore, so it is problematic to compose it with G, online or offline. Just like in static recipe, we remove disambig symbols only on the last stage after composition.

Reply:
How does it affect online composition in terms of search accuracy/speed and fst cache size?

Reply:
We specifically disable cache for disambig removal, so no additional memory is used. As per speed drop, it is maybe 1% or less by measurement, certainly not critical. It is a simple filter anyway.

Overall, this whole thing is not very optimal as I feel it by many different reasons, thought it is a subject of much longer post I'll write soon.

Reply:
Thanks for the response, will look forward for the post!

Reply:
> Overall, this whole thing is not very optimal as I feel it by many different reasons, thought it is a subject of much longer post I'll write soon.

I'm interested in a write up on this, are you still planning to do it?

Reply:
Thanks guys, I finally wrote it here https://alphacephei.com/nsh/2020/03/27/lookahead.html